src/bin/parser-test.rs

```diff
@@ -79,189 +76,112 @@ fn run_tree_test(test_idx: usize, test: &Test, all_results: &mut TestResults) {
-                println!("🌳 Generated tree: ");
-                println!("{}", result.actual_document);
-                println!("----------------------------------------");
-                println!("🌳 Expected tree: ");
-                for line in &test.document {
-                    println!("{line}");
-                }
+                #[cfg(feature = "debug_parser")]
+                println!("❌ {} (wanted: {})", entry.actual, entry.expected);
             }
+            _ => {}
         }
-
-        #[cfg(feature = "debug_parser")]
-        println!("----------------------------------------");
     }
-}
-
-#[allow(dead_code)]
-fn print_test_result(result: &TestResult) {
-    // We need a better tree match system. Right now we match the tree based on the (debug) output
-    // of the tree. Instead, we should generate a document-tree from the expected output and compare
-    // it against the current generated tree.
-    print_node_result(&result.root)
-}
-
-#[allow(dead_code)]
-fn print_node_result(result: &SubtreeResult) {
-    match &result.node {
-        Some(NodeResult::ElementMatchSuccess { actual }) => {
-            println!("✅  {actual}");
-        }
-
-        Some(NodeResult::AttributeMatchFailure { name, expected, .. }) => {
-            println!("❌ {expected}, Found unexpected attribute: {name}");
-        }
-
-        Some(NodeResult::ElementMatchFailure {
-            actual, expected, ..
-        }) => {
-            println!("❌ {expected}, Found unexpected element node: {actual}");
-        }
 
-        Some(NodeResult::TextMatchSuccess { expected }) => {
-            println!("✅  {expected}");
-        }
-
-        Some(NodeResult::TextMatchFailure { expected, text, .. }) => {
-            println!("❌ {expected}, Found unexpected text node: {text}");
-        }
+    for entry in &result.error_results {
+        all_results.assertions += 1;
 
-        Some(NodeResult::DocTypeMatchFailure {
-            actual, expected, ..
-        }) => {
-            println!("❌ {actual}, Found unexpected doctype node: {expected}");
-        }
+        match entry.result {
+            ResultStatus::Success => {
+                all_results.succeeded += 1;
 
-        Some(NodeResult::CommentMatchFailure {
-            expected, comment, ..
-        }) => {
-            println!("❌ {expected}, Found unexpected comment node: {comment}");
+                #[cfg(feature = "debug_parser")]
+                println!(
+                    "✅  ({}:{}) {}",
+                    entry.actual.line, entry.actual.col, entry.actual.message
+                );
+            }
+            ResultStatus::Missing => {
+                all_results.failed += 1;
+
+                #[cfg(feature = "debug_parser")]
+                println!(
+                    "❌ ({}:{}) {} (missing)",
+                    entry.expected.line, entry.expected.col, entry.expected.message
+                );
+            }
```

Comment:
Switching from an enum that had the actual and expected to a bare enum status without fields that requires you to in for what was different feels like slight a step backwards.  But it also seems harmless and easy to revisit.

Reply:
This is one of the things that I'm trying to get used to. So basically you are opting for each result status to be a struct by itself which contains all the information needed for that particular status, instead of having a status enum and accompaning variables?

I'm not 100% sure how this would look like, so maybe you can make a PR for this?